src/transaction/atomic_replication_helper.cpp

```diff
@@ -713,9 +713,10 @@ namespace cublog
 	      }
 	  }
 	// part of scenario (3)
-	// atomic replication sequence within a postpone sequence
+	// atomic replication sequence within a postpone sequence or logical compensation sequence
 	else if (LOG_SYSOP_END == last_entry.m_rectype
-		 && LOG_SYSOP_END_LOGICAL_RUN_POSTPONE == last_entry.m_sysop_end_type)
+		 && (LOG_SYSOP_END_LOGICAL_RUN_POSTPONE == last_entry.m_sysop_end_type
+		     || LOG_SYSOP_END_LOGICAL_COMPENSATE == last_entry.m_sysop_end_type))
```

Comment:
To be honest, I'm having hard time to understand this function. Could you explain some?
1. Why `initial_log_vec_size` must be lesser than 4 if there is no change?
```
 802         // TODO: idle stop condition should normally be removed
 803         if (initial_log_vec_size == m_log_vec.size ())
 804           {
 805             if (initial_log_vec_size >= 4)
 806               {
 807                 if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
 808                   {
 809                     dump ("sequence::can_purge - too many log entries");
 810                   }
 811
 812                 assert (false);
 813               }
 814             break;
 815           }
```
2. Why can we deal with `LOG_SYSOP_END_LOGICAL_COMPENSATE` in the same way with the`LOG_SYSOP_END_LOGICAL_RUN_POSTPONE`?

Reply:
If you think it's hard to explain or the background is too huge, I think I can skip this and just approve it leaving the review to @cristiarg who is the writer of this code.

Reply:
I agree...

1. As far as I understand, currently there are no cases where more than 4 control logs are stacked. That's why the threshold value is set to 4. For example, situations like `LOG_SYSOP_ATOMIC_START`,  `LOG_SYSOP_ATOMIC_START`, `LOG_SYSOP_ATOMIC_START`, `LOG_SYSOP_ATOMIC_START` .. where 4 sysops are nested have been deemed non-existent.
We need @cristiarg opinion.

2. Both 'LOG_SYSOP_END_LOGICAL_COMPENSATE' and 'LOG_SYSOP_END_LOGICAL_RUN_POSTPONE' have `last parent_lsa`. `Last parent lsa` is the LSA value of the log just before the sysop start. Therefore, the action of erasing logs from `last parent lsa +1` to `sysop end` is the same for both logs. 
For 'LOG_SYSOP_END_LOGICAL_UNDO,' we classified it separately because the `last_parent_lsa` value can be null."

Reply:
> Last parent lsa is the LSA value of the log just before the sysop start

Based on the your comment, Can 'last_but_one_entry.m_lsa >= last_entry.m_sysop_end_last_parent_lsa' condition change to 'last_but_one_entry.m_lsa > last_entry.m_sysop_end_last_parent_lsa' in the below?

Reply:
@hornetmj 
Yes, that's correct. In fact, those two values cannot be equal. Based on the current implementation, the LSA of the log before the LOG_SYSOP_ATOMIC_START log is the last parent LSA.  
* last_but_one_entry is LOG_SYSOP_ATOMIC_START
* log_sysop_start_atomic(): set the last parent LSA and then leave the LOG_SYSOP_ATOMIC_START log

In the current atomic replication helper code and related unit tests, it is assumed that the LSA of LOG_SYSOP_ATOMIC_START and the last parent LSA can be the same.

There should not be any issues in terms of functionality because the LOG_SYSOP_ATOMIC_START log only provides timing information.

However, the intent seems important. 

If we want LOG_SYSOP_ATOMIC_START to be equal to the last parent LSA, 
we'll need to change the sequence of logging for LOG_SYSOP_ATOMIC_START. (logging first and set last parent lsa)

If we want to handle LOG_SYSOP_ATOMIC_START > last parent LSA, 
we'll need to modify the atomic replication code and unit tests. 
This is because the unit tests set LOG_SYSOP_ATOMIC_START == last parent LSA.

@cristiarg What are your thoughts on this?

Reply:
> As far as I understand, currently there are no cases where more than 4 control logs are stacked. That's why the threshold value is set to 4. For example, situations like LOG_SYSOP_ATOMIC_START, LOG_SYSOP_ATOMIC_START, LOG_SYSOP_ATOMIC_START, LOG_SYSOP_ATOMIC_START .. where 4 sysops are nested have been deemed non-existent.
We need @cristiarg opinion.

I have no better explanation that this one . :)

Reply:
> If we want LOG_SYSOP_ATOMIC_START to be equal to the last parent LSA,
we'll need to change the sequence of logging for LOG_SYSOP_ATOMIC_START. (logging first and set last parent lsa)
> If we want to handle LOG_SYSOP_ATOMIC_START > last parent LSA,
we'll need to modify the atomic replication code and unit tests.
This is because the unit tests set LOG_SYSOP_ATOMIC_START == last parent LSA.
> @cristiarg What are your thoughts on this?

Is it resolved? Which are we going to change? the unit test or the order of operations?

Reply:
Not resolved yet. I will ask him again

Reply:
@hornetmj  @joohok @wrlawodms 

> Based on the your comment, Can 'last_but_one_entry.m_lsa >= last_entry.m_sysop_end_last_parent_lsa' condition change to 'last_but_one_entry.m_lsa > last_entry.m_sysop_end_last_parent_lsa' in the below?

I understand your concern but I cannot remember why it's `>=` instead of `>`.
All the relevant scenarios have been documented in the header comment. Unfortunately I did not also add log dump snippets - which would have been helpful to discern as to where the `>=` is actually `==`.

Maybe it has to do with repeated occurences of `LOG_SYSOP_END` - see for example scenario (4) in the comment - `LOG_SYSOP_ATOMIC_START` - `LOG_SYSOP_END` - `LOG_SYSOP_END` - and see the comment about the optimization. In this case:
- `last_but_one_entry` will be a `LOG_SYSOP_END`
- and `last_entry` will also be a `LOG_SYSOP_END`

I have nothing against changing the code towards the stricter `>`. And if it crashes, we'll find out.
The problem is that if it crashes - for example - in the tests, I am not sure it will be easy to repro.

And this leads me to suggest that we should add some systematic testing for this atomic replication mechanism - at some point in the future.

Let me know if this answers your questions. Thanks.

Reply:
In the scenario you described, I understand the situation if the sysop is in a consecutive state.
```
LOG_START_ATOMIC_REPL
    LOG_SYSOP_ATOMIC_START
     .. redo records ..
    LOG_SYSOP_END (with LOG_SYSOP_END_LOGICAL_UNDO )
    (sysop start)
     .. redo records ..
    LOG_SYSOP_END (with LOG_SYSOP_END_COMMIT)
LOG_END_ATOMIC_REPL
```

In fact, the question was about which option would be better among the following choices:
1) Should we change the logging method for LOG_SYSOP_ATOMIC_START to align with the logic checking the atomic log sequence?
   --> Modification needed in log_sysop_start_atomic()
2) or should we adjust the logic checking the atomic log sequence to fit the current logging method?
   --> Unit test modification needed

However, even if we choose 1 or 2, it seems to go beyond the scope of this issue. For now, I suggest wrapping up by just adding case following the existing implementation and creating a new issue to address this part.

If we only add case, I plan to add unit tests and conclude the issue.

Reply:
http://jira.cubrid.org/browse/LETS-889